src/meta-srv/src/handler/collect_frontend_cluster_info_handler.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use api::v1::meta::{HeartbeatRequest, Role};
+use common_meta::cluster;
+use common_meta::cluster::{FrontendStatus, NodeInfo, NodeInfoKey, NodeStatus};
+use common_meta::rpc::store::PutRequest;
+use snafu::ResultExt;
+
+use crate::error::{Result, SaveClusterInfoSnafu};
+use crate::handler::{HandleControl, HeartbeatAccumulator, HeartbeatHandler};
+use crate::metasrv::Context;
+
+pub struct CollectFrontendClusterInfoHandler;
+
+#[async_trait::async_trait]
+impl HeartbeatHandler for CollectFrontendClusterInfoHandler {
+    fn is_acceptable(&self, role: Role) -> bool {
+        role == Role::Frontend
+    }
+
+    async fn handle(
+        &self,
+        req: &HeartbeatRequest,
+        ctx: &mut Context,
+        _acc: &mut HeartbeatAccumulator,
+    ) -> Result<HandleControl> {
+        let HeartbeatRequest { header, peer, .. } = req;
+        let Some(header) = &header else {
+            return Ok(HandleControl::Continue);
+        };
+        let Some(peer) = &peer else {
+            return Ok(HandleControl::Continue);
+        };
+
+        let key = NodeInfoKey {
+            cluster_id: header.cluster_id,
+            role: cluster::Role::Frontend,
+            node_id: peer.id,
+        };
+
+        let value = NodeInfo {
+            peer: peer.clone().into(),
+            last_activity_ts: common_time::util::current_time_millis(),
```

Comment:
Why not use the timestamp millis in heartbeat, like in the `CollectDatanodeClusterInfoHandler`'s do?

Reply:
The heartbeat from the frontend is empty, there is no timestamp.

Reply:
I think it's good to have this timestamp in heartbeat. We can create a new good first issue for this.